Automattic-Tracks-iOS/Crash Logging/CrashLogging.swift

```diff
@@ -2,178 +2,251 @@ import Foundation
-        sharedInstance.dataProvider?.didLogErrorCallback?(event)
+        guard let requestBody = try? serializer.serialize() else {
+            DDLogError("⛔️ Unable to send errors to Sentry – error could not be serialized. Attempting to schedule delivery for another time.")
+            errors.forEach {
+                SentrySDK.capture(error: $0)
+            }
+            return
+        }
+
+        let dsn = try SentryDsn(string: dataProvider.sentryDSN)
+        guard let authString = dsn.getAuthString() else {
+            throw Errors.unableToConstructAuthStringError
+        }
+
+        var request = URLRequest(url: dsn.getEnvelopeEndpoint())
+        request.httpMethod = "POST"
+        request.httpBody = requestBody
+        request.addValue(authString, forHTTPHeaderField: "X-Sentry-Auth")
+
+        URLSession.shared.dataTask(with: request) { (responseData, urlResponse, error) in
+            if let error = error {
+                callback(.failure(error))
+                return
+            }
+
+            let didSucceed = 200...299 ~= (urlResponse as! HTTPURLResponse).statusCode
+            callback(.success(didSucceed))
+        }.resume()
     }
 
     /**
-     Writes a message to the Crash Logging system, and includes a stack trace.
+     Writes the error to the Crash Logging system, and includes a stack trace. This method will block the thread until the event is fired.
+
      - Parameters:
-     - message: The message
-     - properties: A dictionary containing additional information about this error
+     - error: The error object
+     - userInfo: A dictionary containing additional data about this error.
      - level: The level of severity to report in Sentry (`.error` by default)
     */
-    static func logMessage(_ message: String, properties: [String: Any]? = nil, level: SentrySeverity = .info) {
+    func logErrorAndWait(_ error: Error, userInfo: [String: Any]? = nil, level: SentryLevel = .error) throws {
+        let semaphore = DispatchSemaphore(value: 0)
 
-        let event = Event(level: level)
-        event.message = message
-        event.extra = properties
-        event.timestamp = Date()
+        var networkError: Error?
+
+        try logErrorImmediately(error, userInfo: userInfo, level: level) { result in
+
+            switch result {
+                case .success:
+                    DDLogDebug("💥 Successfully transmitted crash data")
+                case .failure(let err):
+                    networkError = err
+            }
+
+            semaphore.signal()
+        }
+
+        semaphore.wait()
+
+        if let networkError = networkError {
+            throw networkError
+        }
+    }
 
-        Client.shared?.snapshotStacktrace {
-            Client.shared?.appendStacktrace(to: event)
+    /// A wrapper around the `SentryClient` shim – keeps each layer clean by avoiding optionality
+    private func tryAddingStackTrace(to event: Event) -> Event {
```

Comment:
Nitpick. I found it a bit odd to read "try" as part of the method name.

Having said that, I don't have a solid alternative to suggest. 🤔 

Maybe `addStackTraceWhenPossible(to event:)`?

Reply:
I can see where you're coming from here – I just went with `addStackTrace(to event:)`. 

For the most part this shouldn't fail – the main failure mode would be if the API changed out from underneath us. 

Updated in 20af629f562ad08ff3c937a810a237b868e95a72